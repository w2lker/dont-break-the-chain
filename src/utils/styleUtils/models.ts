export type cssAttribute = 'unset' | 'null' | number | string | undefined | null;

type TArrayWithLength<TItem, TLength extends number> = [TItem, ...TItem[]] & { length: TLength };
export type BoundAttributeArray = TArrayWithLength<string, 1| 2 | 3 | 4>;
