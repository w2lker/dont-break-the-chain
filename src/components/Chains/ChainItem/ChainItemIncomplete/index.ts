import ChainItemIncomplete from './ChainItemIncomplete';
export default ChainItemIncomplete;
