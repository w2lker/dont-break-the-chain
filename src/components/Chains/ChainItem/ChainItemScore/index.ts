import ChainItemScore from './ChainItemScore';
export default ChainItemScore;
