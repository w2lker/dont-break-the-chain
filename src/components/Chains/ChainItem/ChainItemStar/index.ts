import ChainItemStar from "./ChainItemStar";
export default ChainItemStar;