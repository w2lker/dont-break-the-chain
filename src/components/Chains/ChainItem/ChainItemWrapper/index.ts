import ChainItemWrapper from "./ChainItemWrapper";
export default ChainItemWrapper;