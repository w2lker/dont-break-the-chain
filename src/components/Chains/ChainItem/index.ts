import ChainItem from "./ChainItem";
export default ChainItem;