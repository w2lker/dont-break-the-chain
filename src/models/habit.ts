import {List} from 'immutable';
import {colorsKeys} from "../contants/colors";

export const habitStatus = {
  incomplete: 0,
  complete: 1,
  paused: 2,
  start: 3,
  end: 4,
  incompleteToday: 5
};

export interface IHabitDate {
  date?: any;
  status?: string;
  score?: number;
}

export interface IHabit {
  id?: number;
  name?: string;
  color?: colorsKeys;
  icon?: string;
  currentChain?: number;
  longestChain?: number;
  dates?: List<IHabitDate>;
}