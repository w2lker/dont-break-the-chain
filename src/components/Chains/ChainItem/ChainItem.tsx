import * as React from 'react';
import classNames from 'classnames';

import ChainItemWrapper from "./ChainItemWrapper";
import ChainItemScore from "./ChainItemScore";
import ChainItemStar from "./ChainItemStar";
import ChainItemIncomplete from "./ChainItemIncomplete";

import {colorsKeys} from "../../../contants/colors";
import {habitStatusesType} from "../../../models/habit";

export interface IChainItemProps {
  status: habitStatusesType,
  color: colorsKeys,
  value: number,
  onAdd: () => any
}
function ChainItem(props: IChainItemProps) {
  const {status, color, value, onAdd} = props;
  return (
    <ChainItemWrapper
      status={status}
      color={color}
    >
      <ChainItemScore
        status={status}
        color={color}
        value={value}
      />
      <ChainItemStar
        value={value}
        color={color}
      />
      <ChainItemIncomplete
        status={status}
        color={color}
        onAdd={onAdd}
      />
    </ChainItemWrapper>
  );
}

export default ChainItem;