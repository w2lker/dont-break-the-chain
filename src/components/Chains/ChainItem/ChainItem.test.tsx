import { mount, render, shallow } from 'enzyme';
import React from 'react';

import ChainItem, { IChainItemProps } from './ChainItem';

import { habitStatus } from '../../../models/habit';

describe('Chain Component', () => {
  describe('Chain Component statuses snapshots', () => {
    const sampleProps: IChainItemProps = {
      // @ts-ignore
      status: 0,
      color: 'green',
      value: 6,
      onAdd: jest.fn(),
    };

    it('match snapshot statuses', () => {
      const statuses = Object.values(habitStatus);
      statuses.forEach(status => {
        const props = {
          ...sampleProps,
          // @ts-ignore
          status,
        };
        // @ts-ignore
        const component = mount(<ChainItem {...props} />);
        expect(component.debug()).toMatchSnapshot();
      });
    });
    // @ts-ignore
    const getClasses = (instance, componentName) => instance.find(componentName).props().classes;

    function testSingleStatus(status: string) {
      const props: IChainItemProps = {
        ...sampleProps,
        // @ts-ignore
        status: habitStatus[status],
      };
      const component = mount(<ChainItem {...props}/>);
      const a = 3;
      expect(component.find('.habit-chain-wrapper').hasClass(status)).toBeTruthy();
      expect(component.find('.triangleLeft').exists()).toBeFalsy();
      const scoreItemClasses = getClasses(component, 'ChainItemScore');
      expect(component.find(`.${scoreItemClasses.root}`).hasClass('hidden')).toBeFalsy();
      expect(component.find('.mdi-star').exists()).toBeTruthy();
      expect(component.find('.mdi-plus-circle').exists()).toBeFalsy();
    }

    it('renders incomplete habit correctly', () => {
      const incompleteProps: IChainItemProps = {
        ...sampleProps,
          // @ts-ignore
        status: habitStatus.incomplete,
        value: 0,
      };
      const component = mount(<ChainItem {...incompleteProps}/>);
      expect(component.find('.habit-chain-wrapper').hasClass('incomplete')).toBeTruthy();
      expect(component.find('.triangleLeft').exists()).toBeFalsy();
      const scoreItemClasses = getClasses(component, 'ChainItemScore');
      expect(component.find(`.${scoreItemClasses.root}`).hasClass('hidden')).toBeTruthy();
      expect(component.find('.mdi-star').exists()).toBeFalsy();
      expect(component.find('.mdi-plus-circle').exists()).toBeFalsy();
    });
    it('renders complete habit correctly', () => testSingleStatus('complete'));
    it('renders paused habit correctly', () => testSingleStatus('paused'));
    it('renders started habit correctly', () => testSingleStatus('start'));
    it('renders ended habit correctly', () => testSingleStatus('end'));
    it('renders incomplete today habit correctly', () => {
      const props: IChainItemProps = {
        ...sampleProps,
        // @ts-ignore
        status: habitStatus.incompleteToday,
        value: 0,
      };
      const component = mount(<ChainItem {...props}/>);
      expect(component.find('.habit-chain-wrapper').hasClass('incomplete-today')).toBeTruthy();
      expect(component.find('.triangleLeft').exists()).toBeTruthy();
      const scoreItemClasses = getClasses(component, 'ChainItemScore');
      expect(component.find(`.${scoreItemClasses.root}`).hasClass('hidden')).toBeTruthy();
      expect(component.find('.mdi-star').exists()).toBeFalsy();
      expect(component.find('.mdi-plus-circle').exists()).toBeTruthy();
    });
  });
});
