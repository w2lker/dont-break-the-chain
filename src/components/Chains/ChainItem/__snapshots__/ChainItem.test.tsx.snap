// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chain Component Chain Component statuses snapshots match snapshot statuses 1`] = `
"<ChainItem status=\\"habitStatus/incomplete\\" color=\\"green\\" value={6} onAdd={[Function: onAdd]}>
  <ChainItemWrapper status=\\"habitStatus/incomplete\\" color=\\"green\\">
    <div className=\\"habit-chain-wrapper incomplete\\" style={{...}}>
      <ChainItemScore status=\\"habitStatus/incomplete\\" color=\\"green\\" value={6}>
        <span className=\\"score-wrapper hidden bright\\" style={{...}}>
          +6
        </span>
      </ChainItemScore>
      <ChainItemStar value={6} color=\\"green\\">
        <MaterialDesignIcon name=\\"star\\" style={{...}}>
          <span className=\\"material-icon mdi mdi-star\\" style={{...}} />
        </MaterialDesignIcon>
      </ChainItemStar>
      <ChainItemIncomplete status=\\"habitStatus/incomplete\\" color=\\"green\\" onAdd={[Function: onAdd]} />
    </div>
  </ChainItemWrapper>
</ChainItem>"
`;

exports[`Chain Component Chain Component statuses snapshots match snapshot statuses 2`] = `
"<ChainItem status=\\"habitStatus/completed\\" color=\\"green\\" value={6} onAdd={[Function: onAdd]}>
  <ChainItemWrapper status=\\"habitStatus/completed\\" color=\\"green\\">
    <div className=\\"habit-chain-wrapper complete\\" style={{...}}>
      <ChainItemScore status=\\"habitStatus/completed\\" color=\\"green\\" value={6}>
        <span className=\\"score-wrapper bright\\" style={{...}}>
          +6
        </span>
      </ChainItemScore>
      <ChainItemStar value={6} color=\\"green\\">
        <MaterialDesignIcon name=\\"star\\" style={{...}}>
          <span className=\\"material-icon mdi mdi-star\\" style={{...}} />
        </MaterialDesignIcon>
      </ChainItemStar>
      <ChainItemIncomplete status=\\"habitStatus/completed\\" color=\\"green\\" onAdd={[Function: onAdd]} />
    </div>
  </ChainItemWrapper>
</ChainItem>"
`;

exports[`Chain Component Chain Component statuses snapshots match snapshot statuses 3`] = `
"<ChainItem status=\\"habitStatus/paused\\" color=\\"green\\" value={6} onAdd={[Function: onAdd]}>
  <ChainItemWrapper status=\\"habitStatus/paused\\" color=\\"green\\">
    <div className=\\"habit-chain-wrapper paused\\" style={{...}}>
      <ChainItemScore status=\\"habitStatus/paused\\" color=\\"green\\" value={6}>
        <span className=\\"score-wrapper bright\\" style={{...}}>
          +6
        </span>
      </ChainItemScore>
      <ChainItemStar value={6} color=\\"green\\">
        <MaterialDesignIcon name=\\"star\\" style={{...}}>
          <span className=\\"material-icon mdi mdi-star\\" style={{...}} />
        </MaterialDesignIcon>
      </ChainItemStar>
      <ChainItemIncomplete status=\\"habitStatus/paused\\" color=\\"green\\" onAdd={[Function: onAdd]} />
    </div>
  </ChainItemWrapper>
</ChainItem>"
`;

exports[`Chain Component Chain Component statuses snapshots match snapshot statuses 4`] = `
"<ChainItem status=\\"habitStatus/started\\" color=\\"green\\" value={6} onAdd={[Function: onAdd]}>
  <ChainItemWrapper status=\\"habitStatus/started\\" color=\\"green\\">
    <div className=\\"habit-chain-wrapper start\\" style={{...}}>
      <ChainItemScore status=\\"habitStatus/started\\" color=\\"green\\" value={6}>
        <span className=\\"score-wrapper bright\\" style={{...}}>
          +6
        </span>
      </ChainItemScore>
      <ChainItemStar value={6} color=\\"green\\">
        <MaterialDesignIcon name=\\"star\\" style={{...}}>
          <span className=\\"material-icon mdi mdi-star\\" style={{...}} />
        </MaterialDesignIcon>
      </ChainItemStar>
      <ChainItemIncomplete status=\\"habitStatus/started\\" color=\\"green\\" onAdd={[Function: onAdd]} />
    </div>
  </ChainItemWrapper>
</ChainItem>"
`;

exports[`Chain Component Chain Component statuses snapshots match snapshot statuses 5`] = `
"<ChainItem status=\\"habitStatus/ended\\" color=\\"green\\" value={6} onAdd={[Function: onAdd]}>
  <ChainItemWrapper status=\\"habitStatus/ended\\" color=\\"green\\">
    <div className=\\"habit-chain-wrapper end\\" style={{...}}>
      <ChainItemScore status=\\"habitStatus/ended\\" color=\\"green\\" value={6}>
        <span className=\\"score-wrapper bright\\" style={{...}}>
          +6
        </span>
      </ChainItemScore>
      <ChainItemStar value={6} color=\\"green\\">
        <MaterialDesignIcon name=\\"star\\" style={{...}}>
          <span className=\\"material-icon mdi mdi-star\\" style={{...}} />
        </MaterialDesignIcon>
      </ChainItemStar>
      <ChainItemIncomplete status=\\"habitStatus/ended\\" color=\\"green\\" onAdd={[Function: onAdd]} />
    </div>
  </ChainItemWrapper>
</ChainItem>"
`;

exports[`Chain Component Chain Component statuses snapshots match snapshot statuses 6`] = `
"<ChainItem status=\\"habitStatus/incompleteToday\\" color=\\"green\\" value={6} onAdd={[Function: onAdd]}>
  <ChainItemWrapper status=\\"habitStatus/incompleteToday\\" color=\\"green\\">
    <div className=\\"habit-chain-wrapper incomplete-today\\" style={{...}}>
      <div className=\\"custom-background\\" style={{...}} />
      <span className=\\"triangleLeft\\" />
      <span className=\\"triangleRight\\" />
      <ChainItemScore status=\\"habitStatus/incompleteToday\\" color=\\"green\\" value={6}>
        <span className=\\"score-wrapper hidden bright\\" style={{...}}>
          +6
        </span>
      </ChainItemScore>
      <ChainItemStar value={6} color=\\"green\\">
        <MaterialDesignIcon name=\\"star\\" style={{...}}>
          <span className=\\"material-icon mdi mdi-star\\" style={{...}} />
        </MaterialDesignIcon>
      </ChainItemStar>
      <ChainItemIncomplete status=\\"habitStatus/incompleteToday\\" color=\\"green\\" onAdd={[Function: onAdd]}>
        <WithStyles(ForwardRef(IconButton)) onClick={[Function: onAdd]}>
          <ForwardRef(IconButton) classes={{...}} onClick={[Function: onAdd]}>
            <WithStyles(ForwardRef(ButtonBase)) className=\\"MuiIconButton-root\\" centerRipple={true} focusRipple={true} disabled={false} onClick={[Function: onAdd]}>
              <ForwardRef(ButtonBase) classes={{...}} className=\\"MuiIconButton-root\\" centerRipple={true} focusRipple={true} disabled={false} onClick={[Function: onAdd]}>
                <button className=\\"MuiButtonBase-root MuiIconButton-root\\" onBlur={[Function]} onClick={[Function: onAdd]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function]} onMouseLeave={[Function]} onMouseUp={[Function]} onDragLeave={[Function]} onTouchEnd={[Function]} onTouchMove={[Function]} onTouchStart={[Function]} tabIndex={0} type=\\"button\\" disabled={false}>
                  <span className=\\"MuiIconButton-label\\">
                    <MaterialDesignIcon name=\\"plus-circle\\" style={{...}}>
                      <span className=\\"material-icon mdi mdi-plus-circle\\" style={{...}} />
                    </MaterialDesignIcon>
                  </span>
                  <NoSsr>
                    <WithStyles(undefined) center={true}>
                      <ForwardRef(TouchRipple) classes={{...}} center={true}>
                        <span className=\\"MuiTouchRipple-root\\">
                          <TransitionGroup component={{...}} exit={true} childFactory={[Function: childFactory]} />
                        </span>
                      </ForwardRef(TouchRipple)>
                    </WithStyles(undefined)>
                  </NoSsr>
                </button>
              </ForwardRef(ButtonBase)>
            </WithStyles(ForwardRef(ButtonBase))>
          </ForwardRef(IconButton)>
        </WithStyles(ForwardRef(IconButton))>
      </ChainItemIncomplete>
    </div>
  </ChainItemWrapper>
</ChainItem>"
`;
